### PR TITLE
fix: seek subdomain correctly

### PR DESCRIPTION
**Pull request by pauldambra**
We relied on a regex for detecting subdomain in order to choose what cookie domain to set for PostHog storage.

This worked fine for `example.com` or `www.example.com` but it didn't work for `example.co.uk` or `example.com.au` because it would grab the first two parts and so try to set on the public suffix `co.uk` or `com.au`

A lot of suggested solutions on line involve keeping a list of public suffixes, or making DNS lookups to validate values which wouldn't fit here.

But... browsers reject cookies set on public suffixes - despite not offering an API to directly check if something is a public suffix.

So, we can split the current hostname up and then progressively try to set a cookie until the browser accepts it.

For `a.long.domain.someone.is.using.com.au` we would check:

* `.au` rejected by browser
* `.com.au` rejected by browser
* `.using.com.au` accepted by browser

We call this method a lot and the browser can't maintain in-memory variables and navigate between sub-domains, so I also cache the result

I manually tested the code in the developer console of firefox, chrome, and safari

### CODE REVIEW

**Review by benjackwhite — COMMENTED** [review]
Feels like something that needs tests.... can we test it or is it so browser specific that we would essentially be mocking out the whole logic?

File: src/__tests__/storage.test.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            },
+
+            set cookie(value) {
+                //needs to refuse known public suffixes, like a browser would
+                // value arrives like dmn_chk_1699961248575=1;domain=.uk
+                const domain = value.split('domain=')
+                if (['.uk', '.com', '.au', '.com.au', '.co.uk'].includes(domain[1])) return
+                this.value_ += value + ';'
+            },
+        }
```
Comment (pauldambra) [inline src/__tests__/storage.test.ts:20]: @benjackwhite I couldn't get testcafe to test this. But I was having to attach the function to the window and then have testcafe eval it and who knows why it wasn't working.

Sometimes it wasn't present, others it returned the empty string, I think even sometimes it worked, but it was horrible.

So instead I have a fake cookie jar to test the logic with.

And then in the actual code...

File: src/storage.ts [hunk 2]
```diff
@@ -5,7 +5,68 @@ import { DISTINCT_ID, SESSION_ID, SESSION_RECORDING_IS_SAMPLED } from './constan
...
+        // NOTE: Could we use this for cross domain tracking?
+        let matchedSubDomain = seekFirstNonPublicSubDomain(hostname)
+
+        if (!matchedSubDomain) {
+            const originalMatch = originalCookieDomainFn(hostname)
+            if (originalMatch !== matchedSubDomain) {
+                logger.info('Warning: cookie subdomain discovery mismatch', originalMatch, matchedSubDomain)
+            }
+            matchedSubDomain = originalMatch
+        }
```
Comment (pauldambra) [inline src/storage.ts:64]: ... if the new mechanism returns no match we fall back to the original.

Comment (benjackwhite) [reply src/storage.ts:64]: Still somewhat worried about this code but its more of a gut feeling than anything I can actually see wrong with it :D 
Let's try it

**Review by benjackwhite — APPROVED** [review]

**Pull request merged**